lib/modules/asset/AssetService.ts

```diff
@@ -279,239 +279,157 @@ export class AssetService {
     assetsList: string[],
     engineId: string,
     newEngineId: string
-  ): Promise<void> {
-    return lock(`engine:${engineId}:${newEngineId}`, async () => {
-      const recovery = new RecoveryQueue();
+  ): Promise<ApiAssetMigrateTenantResult> {
+    const errors = [];
+    const successes = [];
 
+    //Sanity check
+    if (assetsList.length === 0) {
+      throw new BadRequestError("No assets to migrate");
+    }
+
+    await lock(`engine:${engineId}:${newEngineId}`, async () => {
       if (!user.profileIds.includes("admin")) {
         throw new BadRequestError(
           `User ${user._id} is not authorized to migrate assets`
         );
       }
 
-      try {
-        // check if tenant destination of the the same group
-        const engine = await this.getEngine(engineId);
-        const newEngine = await this.getEngine(newEngineId);
+      // check if tenant destination is in the same group
+      const engine = await this.getEngine(engineId);
+      const newEngine = await this.getEngine(newEngineId);
 
-        if (engine.group !== newEngine.group) {
-          throw new BadRequestError(
-            `Engine ${newEngineId} is not in the same group as ${engineId}`
-          );
-        }
+      if (engine.group !== newEngine.group) {
+        throw new BadRequestError(
+          `Engine ${newEngineId} is not in the same group as ${engineId}`
+        );
+      }
 
-        if (assetsList.length === 0) {
-          throw new BadRequestError("No assets to migrate");
-        }
+      //Get all assets to migrate
+      const assets = await this.sdk.document.mGet<AssetContent>(
+        engineId,
+        InternalCollection.ASSETS,
+        assetsList
+      );
+      errors.push(...assets.errors);
```

Comment:
FYI, doing a push with a spread operator `...` on an array is really bad, the reason for this is when you're using the spread operator like in the following example

```js
x=[];
y=[1,2,3,4];

x.push(...y);
```

It's the equivalent on creating a `push` function that takes 4 parameters, which could be represented like so

```js
x=[]
y=[1,2,3,4]

x.push(y[0], y[1], y[2], y[3]);
```

But here 4 parameters is not going to do any harm but lets say there is 50 elements in the array, now Javascript as to create a new function push at runtime that has to take 50 parameters

```js
x=[]
y=[1,2,3,4,5,...,49,50]

x.push(y[0], y[1], y[2], y[3],...,y[48],y[49]);
```

Now imagine it has to do this each time an array with a different size is used.

This is a not really what is done because V8 does really complex stuff to optimise this, but this give a good perspective to what might happen.

Instead of doing this you should use the `concat` function that takes an array and merge both array into a new one.


Here is a microbenchmark to show you the impact

```
Array concat function x 1,989,683 ops/sec ±4.84% (79 runs sampled)
Spread operator x 195,559 ops/sec ±1.22% (92 runs sampled)
Iterate over array and push element one by one x 241,256 ops/sec ±0.55% (95 runs sampled)
Fastest is Concat
```

This applies for every  instance where you used the spread operator with what might contain lots of data

Reply:
ok